### PR TITLE
Add nested object support in MockQuery

### LINKED ISSUE #81: Nested object support in MockQuery
```
  await fs.collection("tests").add({"testSubObject": {"key": "abc"}});

var result = await fs.collection("tests").where("testSubObject.key", isEqualTo: "abc").getDocuments();

    print(result.documents.length);
```

Right now this returns 0 documents, whilst in FireStore it returns 1 document. 

This has lead to some of our code becoming impossible to test with cloud_firestore_mocks, though I am not sure how many people actually use this feature...

### PR DESCRIPTION
**Pull request by C-Alexander**
Fixes #81

### CODE REVIEW

File: lib/src/mock_query.dart [hunk 1]
```diff
@@ -90,7 +90,7 @@ class MockQuery extends Mock implements Query {
       List<dynamic> whereIn,
       bool isNull}) {
     _QueryOperation operation = (documents) => documents
-        .where((document) => _valueMatchesQuery(document[field],
+        .where((document) => _valueMatchesQuery(_getValueFromDocument(document, field),
```
Comment (atn832) [inline lib/src/mock_query.dart:93]: Your `_getValueFromDocument` utility method is very similar to [util.dart's getSubPath](https://github.com/atn832/cloud_firestore_mocks/blob/45b6837fd614389adf30bf02e4c3b614a1ff1c49/lib/src/util.dart#L3), except that in your case, the separator is `'.'` instead of `'/'`, and yours will throw an exception if the path does not exist.

Instead of making your own method, could you tweak util.dart's getSubPath to accept an optional `separator` argument and reuse it? It'll remove the code duplication and handle missing paths for you.

File: test/mock_query_test.dart [hunk 2]
```diff
@@ -278,6 +278,53 @@ void main() {
...
+        .listen(expectAsync1((QuerySnapshot snapshot) {
+      expect(snapshot.documents.length, equals(1));
+      expect(snapshot.documents.first.data['pager']['number'], equals(10));
+    }));
+
+    instance
+        .collection('users')
+        .document(uid)
+        .collection('bookmarks')
+        .where('pager.number', isGreaterThan: 9)
```
Comment (atn832) [inline test/mock_query_test.dart:321]: Awesome tests! Could you add a final test that checks that for nonexistent paths, it does not throw an exception, but simply returns an empty list? Some thing like:

```dart
 instance
        .collection('users')
        .document(uid)
        .collection('bookmarks')
        .where('pager.some.deep.path', isEqualTo: 10)
        .snapshots()
        .listen(expectAsync1((QuerySnapshot snapshot) {
      expect(snapshot.documents.length, equals(0));
    }));
```

File: lib/src/mock_query.dart (hunk 1 above)
Comment (atn832) [reply lib/src/mock_query.dart:93]: Also in order to use `getSubPath`, you'll probably need to pass `document.data` instead of document and cast the field to `String`: `getSubPath(document.data, field as String)`.

Btw is `Query.where` ever used with a non-string field? What's the expected behavior then?

**Comment by atn832** [comment]
@C-Alexander do you need more guidance on how to polish your PR?

**Comment by psygo** [comment]
This is great. Can we get it to pass? I kinda really need it.

**Comment by atn832** [comment]
Thanks @C-Alexander for filing the bug and suggesting a fix. Eventually someone else submitted #158, so I am closing this PR.
